### PR TITLE
RTDB query gets should not send server requests for listened-to data

### PR DESCRIPTION
**Pull request by IanWyszynski**
b/177263312

### CODE REVIEW

**Comment by IanWyszynski** [comment]
@schmidt-sebastian friendly ping on this, disregard if you were waiting for android to settle before reviewing this one.

File: FirebaseDatabase/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # v7.5.0
 - [added] Implmement `queryStartingAfterValue` and `queryEndingBeforeValue` for FirebaseDatabase query pagination.
 - [added] Added `DatabaseQuery#getData` which returns data from the server when cache is stale (#7110).
+- [added] Check for active-listener in-memory cache values in `FIRDatabaseQuery#getDataWithCompletionBlock` (#7312).
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/CHANGELOG.md:4]: Please use something similar to what you have on Android. It should describe the user-visible behavior change in terms the user understands.

I think you also need to move this to the next release. This unfortunately is now v7.15.1.

File: FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m [hunk 2]
```diff
@@ -4051,14 +4030,35 @@ - (void)testGetQueryBasic {
...
+  FIRDatabase* db = [ref database];
+
+  __block BOOL done = NO;
+
+  @try {
+    [db goOffline];
+
+    [ref getDataWithCompletionBlock:^(NSError* err, FIRDataSnapshot* snapshot) {
+      XCTAssertNotNil(err);
+      //      XCTAssertEqualObjects([err localizedFailureReason], kPersistentConnectionOffline);
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m:4051]: Please comment back in.

File: FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m [hunk 3]
```diff
@@ -4103,35 +4103,20 @@ - (FIRDatabase*)databaseForURL:(NSString*)url name:(NSString*)name {
   }
 }
 
-- (void)testGetRetrievesLatestValueEvenIfCached {
+- (void)testGetSendsServerProbesPersistenceCacheWhenOfflineWithNoListener {
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m:4106]: This test name is quite a handful. Can we make it a bit easier to parse?

File: FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m [hunk 4]
```diff
@@ -4140,13 +4125,47 @@ - (void)testGetRetrievesLatestValueEvenIfCached {
...
+      XCTAssertEqualObjects([err localizedFailureReason], kPersistentConnectionOffline);
+      done = YES;
+    }];
+    WAIT_FOR(done);
+  } @finally {
+    [db goOnline];
+  }
 }
 
 - (void)testGetUpdatesPersistenceCacheWhenEnabled {
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m:4171]: Can you also port the new test from Android?

File: FirebaseDatabase/CHANGELOG.md [hunk 5]
```diff
@@ -1,7 +1,9 @@
+# v7.5.1
+- [added] Optimize `FIRDatabaseQuery#getDataWithCompletionBlock` when in-memory active listener cache exists (#7312).
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/CHANGELOG.md:2]: s/added/changed

File: FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m [hunk 6]
```diff
@@ -4103,7 +4103,7 @@ - (FIRDatabase*)databaseForURL:(NSString*)url name:(NSString*)name {
   }
 }
 
-- (void)testGetSendsServerProbesPersistenceCacheWhenOfflineWithNoListener {
+- (void)testGetReturnsPersistesnceCachedValueWhenOffline {
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m:4106]: Typo in "Persistesnce"

File: FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m [hunk 7]
```diff
@@ -4214,4 +4214,50 @@ - (void)testGetUpdatesPersistenceCacheWhenEnabled {
...
+
+  [readRef removeAllObservers];
+
+  [writeRef setValue:@43
+      withCompletionBlock:^(NSError* error, FIRDatabaseReference* ref) {
+        XCTAssertNil(error);
+        done = YES;
+      }];
+
+  WAIT_FOR(done);
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m:4252]: You need to reset DONE to NO.

**Review by schmidt-sebastian — APPROVED** [review]

**Pull request merged**